# Data Types, Variables, and Operators

**Contents**
1. Key Terms
2. Data Types
3. Operators
4. Variables
5. Operators in Depth
6. Quick Reference

---
# 1. Key Terms

| Term | Definition |
|---|---|
| **State** | Refers to the data that is stored by the program at a specific point in time. |
| **Data Types** | Every value in Python has a type, or category. The type tells Python what the value is and what it can do with it. There are 5 basic types (`str`, `int`, `float`, `bool`, `None`) and 3 types that hold other data or code (`list`, `dict`, and functions). |
| **Operators** | Symbols that generate new data from existing values (e.g., `+`, `>=`, `and`).<br>• **Arithmetic operators (`+`, `-`, `*`, `/`, `//`, `%`, `**`)** calculate a new value from numbers.<br>• **Comparison operators (`==`, `!=`, `<`, `>`, `<=`, `>=`)** compare two values and produce a Boolean.<br>• **Logical operators (`and`, `or`, `not`)** combine or reverse conditions.<br>• **Membership operators (`in`, `not in`)** check whether a value is inside a string, list, or dictionary.<br>• **Identity operators (`is`, `is not`)** check whether two names refer to the very same object.<br>• **Assignment operators (`=`, `+=`, `-=`, and the rest)** store a value in a variable. |
| **Operator precedence** | The order in which Python evaluates the operators in an expression, like multiplication before addition. Parentheses change that order: whatever is inside is evaluated first. It's like PEMDAS, but broader. |
| **Variables** | Named containers for data. This data can change throughout the program by referencing or reassigning it.<br>• A variable is created the first time it's assigned a piece of data.<br>• A name in `ALL_CAPS` is a naming convention that tells others the value is a constant and should never change throughout the program. |
| **`snake_case`** | The way we name functions and variables in Python: lowercase words joined by underscores, like `first_name`. |

---
# 2. Data Types

All values have a **data type** that determines what you can do with that value.

In Python, there are five basic **data types** to start with: `str`, `int`, `float`, `bool`, and `None`.

| Data Type | Syntax Example | What It Represents | Common Use Cases |
|---|---|---|---|
| **`str`** | `"morning"` or `'@abc123'` | Text (any sequence of characters) | Usernames, chat messages, titles, product descriptions, search queries, etc. |
| **`int`** | `100` or `-15` | Whole numbers | Scores in a game, health, a count of lines, a task's position on a list. |
| **`float`** | `10.5` or `0.99` | Numbers with a decimal point. | Prices, averages, percentages, timers. |
| **`bool`** | `True` or `False` | Truth value | Login status, feature toggles (dark mode on/off), win/lose conditions. |
| **`None`** | `None` | Intentional absence of a value | Representing "no result" or "not yet set". |

**Lists**, **dictionaries**, and **functions** are the three types that hold other data or hold code.

| Data Type | Syntax Example | What It Represents | Common Use Cases |
|---|---|---|---|
| **`list`** | `["apple", "banana", "cherry"]` | An ordered list of similar values. | Storing lists (shopping cart items, game high scores, chat messages). |
| **`dict`** | `{"name": "Alex", "age": 25}` | A collection of data relating to one "thing". | Storing user profile data, or data about a product (name, description, price, etc.). |
| **`function`** | `def greet():`<br>`print("Hi!")` | A reusable block of code. | Performing actions, handling events, calculations. |

**Challenge**: Identify the data types of the values below:
```python
[5] # list
"5" # String
5 # Integer
5.0 # Float
{ "value": 5 } # Dictionary
def five(): # Function
    return 5
```

**`type()`** can be used to tell us what **data type** these values are. It is shown below with `print()`.

In [1]:
print(type([5]))             # <class 'list'>
print(type("5"))             # <class 'str'>
print(type(5))               # <class 'int'>
print(type(5.0))             # <class 'float'>
print(type({"value": 5}))    # <class 'dict'>

<class 'list'>
<class 'str'>
<class 'int'>
<class 'float'>
<class 'dict'>


---
# 3. Operators

An **operator** is a symbol like `+`, `*`, `>`, or `==` that takes values and produces a new one. The functionality of the symbol depends on the **data type** of the values you give it.

There's more than one way to store the same information. A light switch could be the text `"on"` or `"off"`, or the Boolean `True` or `False`. This is picked based on what you need to do with it later, since each type works with different operators. Booleans work directly with `if`, `and`, `or`, and `not`. The text version would need comparing every time (`if switch == "on"`).

```python
5 + 2               # 7: + adds numbers
"hello" + " world"  # "hello world": + joins strings
5 * 2               # 10: * multiplies numbers
"hello" * "world"   # TypeError: you can't multiply text by text
```
`+` works on two numbers and on two strings, but it does a different job for each. `*` works on two numbers, but with two strings it means nothing and raises a `TypeError`. That error comes up when an operator is given types it can't handle.

In [2]:
print(10 + 5 * 2)            # 20: multiplication happens before addition
print("B" > "A")             # True: "B" comes after "A". Later letters are greater, and lowercase comes after uppercase
print(2 > 1 and "B" > "A")   # True: both sides are True
# print("5" + 5)           # TypeError: a string and a number can't be added
print(8 / 2)                 # 4.0: / always gives a float
print("3" * 5)               # 33333: a string times a number repeats the string
print(True + True + False)   # 2: True counts as 1, False counts as 0
print([1, 2] * 2)            # [1, 2, 1, 2]: a list times a number repeats the list
print(0 or "Python")         # Python: or skips 0 (counts as false) and gives the other value
print(0.1 + 0.2 == 0.3)      # False: decimals have tiny rounding errors
print([1, 2] == [1, 2])      # True: == checks for the same contents
print([1, 2] is [1, 2])      # False: is checks for the same object, and these are two separate lists

20
True
True
4.0
33333
2
[1, 2, 1, 2]
Python
False
True
False


> **Tip: the Python REPL.** The fastest way to check a prediction is to run `python3` in the Terminal with no file name after it. You get a prompt that evaluates one expression at a time and prints the result. Type `exit()` to leave.

## Resolution Order of Operations

**Operator precedence** is the order in which operators are evaluated. Understanding this is crucial.

In [3]:
fahrenheit = 212
celsius = fahrenheit - 32 * 5 / 9
print(celsius)  # 194.22222222222223

194.22222222222223


We can use this example again. The output should be 100°C.

### Order of Precedence

The **order of precedence** is:

1. Anything inside parentheses `()` is evaluated first, then function calls.
2. Then arithmetic operators: `**` is evaluated before `*`, `/`, `//`, `%`, which are evaluated before `+` and `-`. Operators at the same level are evaluated left to right.
3. Followed by comparison operators (`<`, `<=`, `>`, `>=`, `==`, `!=`).
4. Then logical operators: `not` is evaluated first, then `and`, then `or`.
5. Then conditional expressions (`x if condition else y`) and assignment happen last (`=`).

We use parentheses `()` to take control of the default order, choosing what to evaluate before others.

In [4]:
celsius = (fahrenheit - 32) * 5 / 9
print(celsius)  # 100.0

100.0


**Without Parentheses**:

1. `32 * 5 = 160` because `*` comes before `-`.
2. `160/9 = 17.777...` because `/` is at the same level as `*` and comes next, since it's read left to right.
3. `212 - 17.777... = 194.222...`
4. `celsius = 194.222...` assigns the result.

**With Parentheses**:

1. `(212 - 32) = 180`, since parentheses go first and change the order.
2. `180 * 5 = 900`
3. `900/9 = 100.0`
4. `celsius = 100.0` assigns the result.

The difference is step 1: because we included parentheses, the subtraction goes first, which changes the order of operations.

The same rules apply to comparison and arithmetic in a conditional expression.

In [5]:
temperature = 85
report = "hot" if temperature - 10 > 70 else "mild"
print(report)

hot


**The order of operations here is**:
1. `temperature - 10 = 75`, as arithmetic goes before comparison.
2. `75 > 70 = True`.
3. `"hot" if True else "mild" = "hot"`.
4. `report = "hot"` assigns the result.

In [6]:
# Arithmetic
print(5 + 2 * 3) # 11
print((5 + 2) * 3) # 21
print(2 ** 3 * 2) # 16
print(10 - 2 - 3) # 5

# Comparison
print(5 + 2 > 6) # True
print(5 + (2 > 6)) # 5

# Logical
print(True or False and False) # True: and goes first (False and False is False), then True or False
print((True or False) and False) # False

# Combined
print(5 > 3 and 2 + 2 == 4) # True: True and True
print((5 > 3 and 2 + 2) == 4) # True: (True and 4) gives 4, then 4 == 4
print("even" if 7 % 2 == 0 else "odd") # odd: 7 % 2 is 1, and 1 == 0 is False

11
21
16
5
True
5
True
False
True
True
odd


---
# 4. Variables

A **variable** is a named container or label for data. Variables enable us to perform a series of computations in a program and "save" our progress along the way. In other words, we store our progress in a named container. Being descriptive with the names makes the code easier to read.

These two examples show how variables clean up code and make things easier for us, since we can save expressions in variables to use later on.

```python
print(f"The sum of 4 + 3 + 2 + 1 is {4 + 3 + 2 + 1} and the average is {(4 + 3 + 2 + 1) / 4}")
# Output: The sum of 4 + 3 + 2 + 1 is 10 and the average is 2.5
```
```python
sum = 4 + 3 + 2 + 1
average = sum / 4
message = f"The sum of 4 + 3 + 2 + 1 is {sum} and the average is {average}"
print(message)
# Output: The sum of 4 + 3 + 2 + 1 is 10 and the average is 2.5
```
The first example can be written in one line, but it is a very long line that is difficult to read. It also repeats expressions that could be stored in a variable and **referenced** instead of being written out again.

With **variables**, each expression is stored and labeled nicely for readability. The `sum` variable is made first and then used to calculate `average`. Both are then **referenced** in the `message` variable, which gets printed. More lines are used, but the code is cleaner and easier to follow.

## Using Variables: Assign, Reassign, Reference

Three things we can do with a variable:

1. **Assign**: Creating a new variable by giving it a name and a value ("**assign** the variable a value").
2. **Reassign**: Giving an existing variable a new value to hold ("**reassign** the variable").
3. **Reference**: Using the value held by the variable ("**reference** the variable").

In [7]:
# Create a variable count and assign it the value 0
count = 0

# Reference the count variable to access its current value, 0
print(f"count starts at {count}")

# Reassign the variable to hold a new value: 1
count = 1

# Reference the count variable again to access its current value, 1
print(f"count is now {count}")

count starts at 0
count is now 1


We can also ***reference*** the value of a variable to ***reassign*** it a new value:

In [8]:
# The right side (count + 1) resolves before the assignment. count becomes 2
count = count + 1

# += is an "augmented assignment" operator. It is the same as `count = count + 1`. count becomes 3
count += 1

# It works with any amount. count becomes 6
count += 3

# You can use every other arithmetic operator with =. count becomes 12
count *= 2

print(f"count is now {count}")  # Output: count is now 12

count is now 12


**Predict, then run:** What is the output of the code below?

**My prediction:** This will print `helloworld!helloworld!helloworld!`.

In [9]:
msg = 'hello'
msg += 'world' + '!' * 3
print(msg)

helloworld!!!


My prediction was wrong. I did not take the order of operations into account. I thought it would multiply the complete string 3 times, but it only multiplied the `'!'` 3 times.

1. `'!' * 3` evaluates to `'!!!'`
2. `'world' + '!!!'` evaluates to `'world!!!'`
3. `msg += 'world!!!'` reassigns `msg` to `'helloworld!!!'`

> **Warning:** `+` joins strings exactly as they are. It does not add a space for you, which is why the result is `helloworld!!!` and not `hello world!!!`. Put the space inside one of the strings, like `'hello' + ' world'`.

## Naming Conventions

Variables should always be named using **`snake_case`**: lowercase words joined by underscores. The names should describe the data they hold, for the best readability.

A global variable that we want to keep as a constant, because it doesn't change throughout the program, is written in `ALL_CAPS`. Every Python programmer should read that as "do not reassign me".

In [10]:
# Unclear or hard to read
t = 201                  # what is t?
lastyearstips = 201      # the words run together
lastYearsTips = 201      # camelCase: common in other languages, but not the Python style

# Clear: snake_case that describes the value
last_years_tips = 201

# A constant: ALL_CAPS tells other programmers "do not reassign me"
MAX_HEALTH = 100

print(last_years_tips)
print(MAX_HEALTH)

201
100


`ALL_CAPS` is only a convention. Python will not stop you from reassigning `MAX_HEALTH`, so it is up to the programmer to leave it alone.

---
# 5. Operators in Depth

## Arithmetic Operators

**Arithmetic operators** perform standard mathematical calculations or sequence manipulation like <mark>**string concatenation**</mark> (`"hi" + "world"`) and list repetition. Concatenation has its own section below.

| Operator | Name | What it does | Example | Result |
|---|---|---|---|---|
| `+` | Addition | Adds two numbers | `7 + 2` | `9` |
| `-` | Subtraction | Subtracts the right number from the left | `7 - 2` | `5` |
| `*` | Multiplication | Multiplies two numbers | `7 * 2` | `14` |
| `/` | Division | Divides, and always gives a float | `7 / 2` | `3.5` |
| `//` | Floor division | Divides and rounds down to a whole number | `7 // 2` | `3` |
| `%` | Modulo | Gives the remainder after dividing | `7 % 2` | `1` |
| `**` | Exponent | Raises to a power | `7 ** 2` | `49` |

> **Note on `//`:** the type of the result depends on what you give it. If both numbers are `int`, the result is an `int` (`7 // 2` is `3`). If either number is a `float`, the result is a `float` (`7.0 // 2` is `3.0`). It also rounds **down**, so `-7 // 2` is `-4`, not `-3`.

`+` and `*` also work on strings and lists:

| Example | Result | What happened |
|---|---|---|
| `"hi" + "world"` | `"hiworld"` | `+` joins two strings |
| `"ha" * 3` | `"hahaha"` | `*` repeats a string |
| `[1, 2] * 2` | `[1, 2, 1, 2]` | `*` repeats a list |

```python
5 + 3           # 8
"a" + "b"       # "ab"
10 - 4          # 6
4 * 2.5         # 10.0
"hi" * 3        # "hihihi"
7 / 2           # 3.5
4 / 2           # 2.0
7 // 2          # 3
7.0 // 2        # 3.0
7 % 2           # 1
2 ** 3          # 8
```

### Concatenation

<mark>**Concatenation** means joining two sequences end to end to make one longer sequence.</mark> In Python, the `+` operator does this for strings and for lists.

The word comes from "chain": the pieces are linked together in the order you write them, like links in a chain.

| Example | Result | What happened |
|---|---|---|
| `"hi" + "world"` | `"hiworld"` | Two strings joined into one |
| `"hi" + " " + "world"` | `"hi world"` | A space is a string too, so it can be joined in |
| `[1, 2] + [3, 4]` | `[1, 2, 3, 4]` | Two lists joined into one |
| `"ha" * 3` | `"hahaha"` | **Repetition**, not concatenation: the same string joined to itself |

Rules to remember:

- **Both sides must be the same type.** A string joins with a string, and a list joins with a list.
- **`+` does not add spaces.** The strings are joined exactly as they are.
- **A string and a number cannot be concatenated.** `"Age: " + 25` raises a `TypeError`. Convert the number first with `str(25)`, or use an f-string.
- **The originals are not changed.** Concatenation makes a new string or list.
- **`+=` concatenates and reassigns** in one step: `msg += "!"` is the same as `msg = msg + "!"`.

In [11]:
first_name = "Xavier"
last_name = "Rodriguez"

# Joined exactly as they are: no space is added
print(first_name + last_name)

# Add the space yourself
full_name = first_name + " " + last_name
print(full_name)

# A number must be converted to a string first
age = 25
print(full_name + " is " + str(age))

# An f-string does the same job without + or str()
print(f"{full_name} is {age}")

# Lists concatenate too
print([1, 2] + [3, 4])

# The originals are unchanged
print(first_name)

XavierRodriguez
Xavier Rodriguez
Xavier Rodriguez is 25
Xavier Rodriguez is 25
[1, 2, 3, 4]
Xavier


> **Concatenation vs. f-strings:** both build a longer string. Concatenation with `+` is fine for joining a couple of strings. When numbers or several values are involved, an f-string is shorter and avoids the `TypeError`.

## Comparison (Relational) Operators

Comparison operators compare two values and return a Boolean value depending on the comparison:

| Operator | Name | `True` when... | Example | Result |
|---|---|---|---|---|
| `==` | Equal to | Both values are the same | `5 == 5` | `True` |
| `!=` | Not equal to | The values are different | `5 != 3` | `True` |
| `>` | Greater than | The left is bigger than the right | `5 > 3` | `True` |
| `<` | Less than | The left is smaller than the right | `5 < 3` | `False` |
| `>=` | Greater than or equal to | The left is bigger than or the same as the right | `5 >= 5` | `True` |
| `<=` | Less than or equal to | The left is smaller than or the same as the right | `5 <= 3` | `False` |

Things to watch for:

| Example | Result | What happened |
|---|---|---|
| `x = 5` vs `x == 5` | | One `=` stores a value. Two `==` asks "are these equal?" |
| `5 == "5"` | `False` | A number and a string are never equal, even if they look alike |
| `5 == 5.0` | `True` | An `int` and a `float` with the same value are equal |
| `"B" > "A"` | `True` | Strings compare in alphabetical order |
| `"apple" == "Apple"` | `False` | String comparison is case-sensitive |

```python
5 == 5              # True
"a" != "b"          # True
3 < 5               # True
10 > 2              # True
[1, 2] <= [1, 2]    # True
5.0 >= 5            # True
5 < "hello"         # TypeError
```

## Logical Operators

Logical operators combine Boolean values into a single `True` or `False` value.

| Operator | `True` when... | Example | Result |
|---|---|---|---|
| `and` | **Both** sides are `True` | `True and False` | `False` |
| `or` | **At least one** side is `True` | `True or False` | `True` |
| `not` | The value is `False` (it flips it) | `not True` | `False` |

Every combination, written out:

| `and` needs **both** | Result | | `or` needs **one** | Result |
|---|---|---|---|---|
| `True and True` | `True` | | `True or True` | `True` |
| `True and False` | `False` | | `True or False` | `True` |
| `False and True` | `False` | | `False or True` | `True` |
| `False and False` | `False` | | `False or False` | `False` |

`and` is only `True` when both sides are `True`. `or` is only `False` when both sides are `False`.

Things to watch for:

| Example | Result | What happened |
|---|---|---|
| `5 > 3 and 2 > 1` | `True` | The comparisons are evaluated first, then `and` combines them |
| `True or False and False` | `True` | `and` is evaluated before `or` |
| `(True or False) and False` | `False` | Parentheses change that order |
| `not 5 > 3` | `False` | `5 > 3` is evaluated first, then `not` flips it |
| `0 or "Python"` | `"Python"` | `or` gives back the first value that counts as true |

```python
True and False      # False
True or False       # True
not True            # False
5 > 3 and 2 > 1     # True
```

In chapter 1.4 we will learn that `and` and `or` also accept values that are not Booleans, using a rule called truthiness. That is what the `0 or "Python"` line in the challenge above was doing.

## Membership and Identity Operators

Membership operators (`in`, `not in`) test whether a value exists inside a sequence or container. Identity operators (`is`, `is not`) test whether two names/variables refer to the same object. In other words, `is` gives `True` when two variables point to the same object.

### Membership: `in` and `not in`

| Operator | `True` when... | Example | Result |
|---|---|---|---|
| `in` | The value is found inside | `"a" in "apple"` | `True` |
| `not in` | The value is **not** found inside | `"z" not in "apple"` | `True` |

It works on strings, lists, and dictionaries:

| Example | Result | What happened |
|---|---|---|
| `"app" in "apple"` | `True` | For strings, it checks for a substring |
| `"A" in "apple"` | `False` | String checks are case-sensitive |
| `3 in [1, 2, 3]` | `True` | For lists, it checks for a matching item |
| `"3" in [1, 2, 3]` | `False` | The string `"3"` is not the number `3` |
| `"name" in {"name": "Xavier"}` | `True` | For dictionaries, it checks the keys |
| `"Xavier" in {"name": "Xavier"}` | `False` | It does not check the values, only the keys |

### Identity: `is` and `is not`

| Operator | `True` when... | Example | Result |
|---|---|---|---|
| `is` | Both sides are the **same object** | `mood is None` | `True` if `mood` is `None` |
| `is not` | The two sides are **different objects** | `mood is not None` | `True` if `mood` has a value |

`==` and `is` ask different questions:

| Example | Result | What happened |
|---|---|---|
| `[1, 2] == [1, 2]` | `True` | `==` asks: do they have the same contents? |
| `[1, 2] is [1, 2]` | `False` | `is` asks: are they the same object? These are two separate lists |
| `b = a`, then `a is b` | `True` | `b` is a second name for the same list, not a copy |

Use `==` to compare values. Use `is` mainly for checking `None`.

In [12]:
first_name = [1, 2]
middle_name = [1, 2]
print(first_name is middle_name)

False


Each `[1, 2]` creates a new list, so the two names above point to two separate objects and `is` gives `False`.

### Why `is` can be `True` for two identical strings

With strings, the same test can give `True`:

In [13]:
first_name = "Xavier"
middle_name = "Xavier"

print(first_name is middle_name)

True


This happens because strings are **immutable** and lists are **mutable**.

| Type | Can it change? | Two identical values written separately |
|---|---|---|
| String | No (immutable) | Python may reuse **one** object for both, to save memory |
| List | Yes (mutable) | Always **two** separate objects |

- A string can never change, so it is safe for two names to share one string.
- A list can change. If two names shared one list, changing it through one name would change it for the other. So Python always makes a new list.

Python reusing a string is a shortcut, not a guarantee. Below, both strings hold the same text, but one is built while the program runs, so Python stores it as a separate object:

In [14]:
a = "Xavier"
b = "Xav" + "ier".lower()

print(a)
print(b)
print(a == b)   # same contents
print(a is b)   # not the same object

Xavier
Xavier
True
False


`==` gave the right answer both times. `is` gave `True` in one case and `False` in the other, for the same text.

> **Rule:** always compare strings (and other values) with `==`. Keep `is` for checking `None`.

## Assignment Operators

Assignment operators store a value in a variable. `=` is simple assignment. The others are **augmented assignment** operators: they do the calculation and store the result back in the same variable, in one step.

Each row below starts from `x = 10`:

| Operator | Example | Same as | `x` becomes |
|---|---|---|---|
| `=` | `x = 10` | | `10` |
| `+=` | `x += 5` | `x = x + 5` | `15` |
| `-=` | `x -= 3` | `x = x - 3` | `7` |
| `*=` | `x *= 2` | `x = x * 2` | `20` |
| `/=` | `x /= 4` | `x = x / 4` | `2.5` |
| `//=` | `x //= 3` | `x = x // 3` | `3` |
| `%=` | `x %= 3` | `x = x % 3` | `1` |
| `**=` | `x **= 2` | `x = x ** 2` | `100` |

Things to watch for:

| Example | What happened |
|---|---|
| `x = 5` vs `x == 5` | One `=` stores a value. Two `==` compares two values. |
| `msg += "!"` | `+=` also works on strings (and lists): it adds to the end. |
| `x /= 2` | `/=` always gives a float, like `/` does. |
| `msg += 'world' + '!' * 3` | The whole right side is evaluated first, then the result is stored. |
| `count += 1` before `count` exists | `NameError`: a variable must be assigned before it can be updated. |

In [15]:
health = 100
zombie_damage = 15

health -= zombie_damage    # same as: health = health - zombie_damage
print(health)

health += 10               # picked up a health pack
print(health)

health *= 2                # double health power-up
print(health)

health //= 3               # floor division keeps it a whole number
print(health)

85
95
190
63


## The Conditional Expression

A conditional expression is an `if`/`else` written on one line. It **produces a value**, so it can be used anywhere a value can go: in an assignment, inside `print()`, or inside an f-string.

```python
value_if_true if condition else value_if_false
```

Read it as: "give me the first value **if** the condition is `True`, **else** give me the second value."

In [16]:
health = 85
status = "alive" if health > 0 else "defeated"
print(status)

number = 7
print("even" if number % 2 == 0 else "odd")

print(f"The player is {'healthy' if health >= 50 else 'hurt'}.")

alive
odd
The player is healthy.


The first example does the same job as this longer `if`/`else` statement:

```python
if health > 0:
    status = "alive"
else:
    status = "defeated"
```

Order of evaluation: the condition in the middle is evaluated first, then Python picks one of the two values.

Use a conditional expression for a simple choice between two values. For more than two outcomes, or when each branch needs several lines, use a regular `if` statement.

---
# 6. Quick Reference

**Data types**

| Type | Example | What it is |
|---|---|---|
| `str` | `"hello"` | Text |
| `int` | `25` | Whole number |
| `float` | `4.5` | Number with a decimal point |
| `bool` | `True` | True or false |
| `None` | `None` | No value yet |
| `list` | `[1, 2, 3]` | Several values in order |
| `dict` | `{"name": "Alex"}` | Values with labels (keys) |
| function | `def greet():` | A reusable block of code |

**Operators**

| Group | Operators | What they produce |
|---|---|---|
| Arithmetic | `+` `-` `*` `/` `//` `%` `**` | A number (or a joined / repeated string or list) |
| Comparison | `==` `!=` `<` `>` `<=` `>=` | `True` or `False` |
| Logical | `and` `or` `not` | Combines or flips conditions |
| Membership | `in` `not in` | `True` if a value is (or is not) inside |
| Identity | `is` `is not` | `True` if both names point to the same object |
| Assignment | `=` `+=` `-=` `*=` `/=` `//=` `%=` `**=` | Stores a value in a variable |
| Conditional expression | `a if condition else b` | One of two values |

**Order of precedence** (first to last)

1. Parentheses `()`, then function calls
2. `**`
3. `*` `/` `//` `%`
4. `+` `-`
5. Comparisons: `==` `!=` `<` `>` `<=` `>=`
6. `not`
7. `and`
8. `or`
9. Conditional expression, then assignment `=`

**Variables**

| Action | Example |
|---|---|
| Assign | `count = 0` |
| Reference | `print(count)` |
| Reassign | `count = 1` |
| Reassign using its own value | `count += 1` |
| Name a variable | `last_years_tips` (`snake_case`) |
| Name a constant | `MAX_HEALTH` (`ALL_CAPS`) |

**Worth remembering**

- `/` always gives a float. `//` rounds down to a whole number.
- **Concatenation** is joining strings (or lists) end to end with `+`. It does not add spaces.
- A string and a number can't be added: `"5" + 5` is a `TypeError`.
- `0.1 + 0.2 == 0.3` is `False`, because decimals have tiny rounding errors.
- Compare values with `==`. Keep `is` for checking `None`.
- One `=` assigns. Two `==` compares.